From b72c74df1d84ff60341fafb49bc1a3f36217db56 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Mon, 20 Mar 2017 14:45:06 +0800
Subject: [PATCH 18/45] udc: Add PM support for USB2.0 device driver

- Since U3D and UDC share the same device controller, need to
  suspend/resume together with the U3D driver
- Stop the phy driver in suspend, start again in resume

Change-Id: Ie767eef673f264b82dbb1cc609a43ad72a068487
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37634
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
---
 drivers/usb/gadget/udc/mv_udc_core.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/drivers/usb/gadget/udc/mv_udc_core.c b/drivers/usb/gadget/udc/mv_udc_core.c
index a7b9c06a50cc..563780961140 100644
--- a/drivers/usb/gadget/udc/mv_udc_core.c
+++ b/drivers/usb/gadget/udc/mv_udc_core.c
@@ -2476,6 +2476,12 @@ static int mv_udc_suspend(struct device *dev)
 		mv_udc_disable_internal(udc);
 	}
 
+	/* PHY exit if there is */
+	if (udc->utmi_phy) {
+		phy_power_off(udc->utmi_phy);
+		phy_exit(udc->utmi_phy);
+	}
+
 	return 0;
 }
 
@@ -2490,6 +2496,20 @@ static int mv_udc_resume(struct device *dev)
 	if (udc->transceiver)
 		return 0;
 
+	/* PHY init if there is */
+	if (udc->utmi_phy) {
+		retval = phy_init(udc->utmi_phy);
+		if (retval)
+			return retval;
+
+		retval = phy_power_on(udc->utmi_phy);
+		if (retval) {
+			phy_power_off(udc->utmi_phy);
+			phy_exit(udc->utmi_phy);
+			return retval;
+		}
+	}
+
 	if (!udc->clock_gating) {
 		retval = mv_udc_enable_internal(udc);
 		if (retval)
-- 
2.35.3

